### ISSUE TITLE
Feature request: let us disable the metadata hash

### ISSUE THREAD

**Issue by paulrberg** [ethereum/remix-project#1559]
It seems like Remix includes the metadata hash in the compilation of the contracts. This can lead to various confusing outcomes, such as the one described in https://github.com/ethereum/remix-project/issues/293.

It would be nice if we could disable the metadata hash via a check box. Most development frameworks offer this feature today, e.g. this is how I do it in Hardhat:

https://github.com/paulrberg/solidity-template/blob/2de6b59757e0dd7f467d7549074329391a5d671f/hardhat.config.ts#L83

**Comment by chriseth** [comment]
It's weird because the metadata hash (or at least the metadata) is exactly the feature that allows easy recompilation.

Is the fact that there is no easy way to re-compile something given just the metadata json the problem?

**Comment by paulrberg** [comment]
> Is the fact that there is no easy way to re-compile something given just the metadata json the problem?

As end user we often don't have access to the metadata json. In fact, I would wager that most of don't even know that such file exists (I confess that I didn't know until recently).

We discussed the same issue in more details in this convo re my Solidity template:

https://github.com/paulrberg/solidity-template/issues/31

**Comment by chriseth** [comment]
Yes, I am aware that there is some kind of learning hump we have to overcome, but my hope is that after that, everything is much smoother. If you do not remove the metadata hash and publish the metadata json, then (we still have to talk to tooling people there) you can resolve all the information (available functions, documentation, source, ...) by just supplying the address of a contract. If you delete the metadata hash you have to rely on centralized services like etherscan to provide this for you.

Truffle stores the metadata json in the build artefacts, and while debugging the nondeterminism issue, I noticed that the "build-info" generated by hardhat also looks very close to a metadata file.

But yeah, we will talk to tooling people to make the metadata json more easily available.

**Comment by paulrberg** [comment]
Oh, I thought that relying on centralised services like Etherscan is the only possible way to resolve information about the original source code.

Thanks for explaining this. I will dedicate time to understand how the metadata works.
